### PR TITLE
GEODE-10354: Convert enum like classes to enum.

### PR DESCRIPTION
**Pull request by pivotal-jbarrett**
### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [ ] Is your initial contribution a single, squashed commit?

- [ ] Does `gradlew build` run cleanly?

- [ ] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### CODE REVIEW

**Comment by onichols-pivotal** [comment]
this PR appears to be abandoned, can it be closed?

**Review by onichols-pivotal — CHANGES_REQUESTED** [review]
_comment deleted_

**Review by onichols-pivotal — APPROVED** [review]

**Review by rhoughton-pivot — APPROVED** [review]

**Review by DonalEvans — APPROVED** [review]
Approving for the files I'm codeowner for

**Review by jinmeiliao — APPROVED** [review]

File: geode-dunit/src/main/java/org/apache/geode/cache30/RegionTestCase.java
Comment (demery-pivotal) [inline geode-dunit/src/main/java/org/apache/geode/cache30/RegionTestCase.java:3738]: Can `getMirrorType().isMirrored()` be replaced with information from the `DataPolicy`?

**Review by demery-pivotal — APPROVED** [review]
Changes in `geode-dunit` and `geode-junit` approved.

**Review by kirklund — APPROVED** [review]

File: geode-dunit/src/main/java/org/apache/geode/cache30/RegionTestCase.java
Comment (pivotal-jbarrett) [reply geode-dunit/src/main/java/org/apache/geode/cache30/RegionTestCase.java:3738]: Don't know. It is out of scope for this current change.

File: geode-core/src/test/java/org/apache/geode/internal/cache/tier/sockets/command/Put70Test.java
Comment (agingade) [inline geode-core/src/test/java/org/apache/geode/internal/cache/tier/sockets/command/Put70Test.java:335]: The older code seems to be looking to know if its configured with persistent data-policy.
The new change is different than that. Is it what is expected in the test.
In the previous above change; for the same old code its returning DataPolicy.PERSISTENT_REPLICATE.
Same comment with following two changes in this file.

Comment (gesterzhou) [reply geode-core/src/test/java/org/apache/geode/internal/cache/tier/sockets/command/Put70Test.java:335]: logically it's the same here.

**Review by gesterzhou — APPROVED** [review]

**Review by upthewaterspout — CHANGES_REQUESTED** [review]
I did a quick test, and it looks like this is a breaking change to the serialization format of this class. I think that's why we haven't made this conversion up to this point:

```
Caused by: java.io.InvalidClassException: cannot bind non-enum descriptor to an enum class
	at java.io.ObjectStreamClass.initNonProxy(ObjectStreamClass.java:687)
	at java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:1885)
	at java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1751)
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2042)
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1573)
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:431)
	at org.apache.geode.internal.InternalDataSerializer.readSerializable(InternalDataSerializer.java:2720)
```

**Comment by pivotal-jbarrett** [comment]
> I did a quick test, and it looks like this is a breaking change to the serialization format of this class. I think that's why we haven't made this conversion up to this point:
> 
> ```
> Caused by: java.io.InvalidClassException: cannot bind non-enum descriptor to an enum class
> 	at java.io.ObjectStreamClass.initNonProxy(ObjectStreamClass.java:687)
> 	at java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:1885)
> 	at java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1751)
> 	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2042)
> 	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1573)
> 	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:431)
> 	at org.apache.geode.internal.InternalDataSerializer.readSerializable(InternalDataSerializer.java:2720)
> ```

Which test?

**Comment by upthewaterspout** [comment]
@pivotal-jbarrett - Here's a test you can play with, based on this PR - https://github.com/upthewaterspout/geode/pull/new/feature/test-compatibility-of-enum-DataPolicy

File: geode-core/src/test/java/org/apache/geode/internal/cache/tier/sockets/command/Put70Test.java
Comment (pivotal-jbarrett) [reply geode-core/src/test/java/org/apache/geode/internal/cache/tier/sockets/command/Put70Test.java:335]: You fill find lots of places in the old tests where the combinations of mocks and return values used for `DataPolicy` actually created mutually exclusive combinations but it didn't matter for the sake of the test. For the changes here I tried to match up what the test was actually expecting for the correct combinations that aligned to an enum value.

Comment (agingade) [reply geode-core/src/test/java/org/apache/geode/internal/cache/tier/sockets/command/Put70Test.java:335]: Based on the test name; is it expecting it to treat as persistent when it is accessor but not configured with persistence and has persistent members?
If thats the case; the change seems to be defeating the purpose by not having:
when(dataPolicy.withPersistence()).thenReturn(false);

**Comment by pivotal-jbarrett** [comment]
@upthewaterspout, this brings up an excellent question. Does the use of the `Serializable` interface outside of the direct API, which does not utilize it, constitute part of the binary compatibility contract for a public class. Our API or internals do not transmit the classes in question as serializables but the test framework does. As a result of having the `Serializable` interface on the class are we stuck making sure that external use of this class maintains serialization compatibility? 

If the answer to this is yes, then it will likely put a damper on future mutation of types that either directly or indirectly, like `enum`, implement `Serializable`. There are many issues with versioning `Serializable` classes that may force us to consider `@Deprecated` annotations on these sorts of types and making new types to replace them anytime we want to mutate them.

If the answer is no, then how should we communicate this to users? We could define a `@NotSerializable` annotation and Javadoc tag to hopefully inform the user not to use it in serialization. We could implement the write/readObject methods and throw exceptions to prevent it at runtime.

**Review by onichols-pivotal — CHANGES_REQUESTED** [review]
due to code freeze, please hold off merging this

**Comment by metatype** [comment]
Closing this out until the open questions are resolved.

**Pull request closed without merging**
